SYCL/ESIMD/api/functional/value.hpp

```diff
@@ -0,0 +1,185 @@
+//      testing. ----------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+///
+/// \file
+/// This file provides functions that let obtain data for test according to
+/// current underlying type.
+///
+//===----------------------------------------------------------------------===//
+
+#pragma once
+#include "type_traits.hpp"
+#include <CL/sycl.hpp>
+
+#include <climits>
+#include <limits>
+#include <type_traits>
+
+namespace esimd_test {
+namespace api {
+namespace functional {
+
+// Analog to C++20 bit_cast
+template <typename To, typename From,
+          typename std::enable_if<(sizeof(To) == sizeof(From)) &&
+                                      std::is_trivially_copyable_v<From> &&
+                                      std::is_trivial_v<To>,
+                                  int>::type = 0>
+To bit_cast(const From &src) noexcept {
+  To dst;
+  std::memcpy(&dst, &src, sizeof(To));
+  return dst;
+}
+
+// Initializes the sycl::half value by using two bytes given:
+//  - the higher byte, including the sign bit
+//  - the lower byte, including the part of mantissa
+//
+// This implementation doesn't depend on:
+//  - the byte order of both the unsigned types and the floating type itself;
+//  - the existence of the optional std::uint16_t type
+//  - compiler optimisations related to the strict aliasing rules
+sycl::half half_from_bytes(unsigned char hi, unsigned char lo) noexcept {
+  const size_t size = sizeof(sycl::half);
+  static_assert(CHAR_BIT == 8, "Unexpected byte size, input values may broke");
+  static_assert(size == 2, "Invalid number of bytes for half type");
+
+  const unsigned char in[size] = {lo, hi};
+  unsigned char index[size];
+  unsigned char out[size];
+
+  sycl::half indexHint = 2;
+  const unsigned char indexCoeff = 64;
+
+  memcpy(index, &indexHint, size);
+  index[0] /= indexCoeff;
+  index[1] /= indexCoeff;
+
+  assert(index[0] + index[1] == 1);
+
+  out[0] = in[index[0]];
+  out[1] = in[index[1]];
+  return esimd_test::bit_cast<sycl::half>(out);
+}
+
+// Struct that provise methods for constructing inf, lowest, nan, denorm_min and
+// max values for reference data.
+template <typename DataT> struct value {
+  static DataT inf() {
+    static_assert(type_traits::is_floating_point_v<DataT>,
+                  "Infinity required only for floating point data types.");
+
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return half_from_bytes(0b01111100u, 0b00000000u);
+    } else {
+      return std::numeric_limits<DataT>::infinity();
+    }
+  }
+
+  static DataT lowest() {
+    if constexpr (std::is_same_v<DataT, sycl::half>) {
+      return -half_from_bytes(0b01111011u, 0b11111111u);
```

Comment:
Either
```suggestion
      return -max();
```
or
```suggestion
      return half_from_bytes(0b1111011u, 0b11111111u);
```
The first one is the preferred one - because we don't duplicate the actual bits.

Reply:
this comment was applied in this commit: 4bd389b51699b9b2e1d891e8bcab1c536cd45a90